### ISSUE TITLE
Add interval: "batch" option for learning rate schedulers

### ISSUE THREAD

**Issue by smithblack-0** [Lightning-AI/pytorch-lightning#21491]
### Description & Motivation

# Feature Request

Currently, PyTorch Lightning only supports `interval: "step"` (per optimizer step) and `interval: "epoch"` (per epoch) for learning rate schedulers. This limitation creates problems when using gradient accumulation with varying accumulation factors during training.

# Motivation

Despite Lightning's Trainer explicitly supporting adaptive `accumulate_grad_batches`, there is no way to schedule learning rates or other hyperparameters per batch. This makes runs with varying accumulation difficult to compare, as they should be scheduled according to tokens/batches seen rather than optimizer steps taken. This also makes investigating training dynamics and batch-size-dependent effects more difficult, as research requires per-batch granularity rather than per-optimizer-step granularity. 

### Pitch

# Proposal: 'batch' interval

Add `interval: "batch"` as a supported option in `configure_optimizers()` when defining schedules under the "interval" argument.

```python
    def configure_optimizers(self):
        return {
            "optimizer": optimizer,
            "lr_scheduler": {
                "scheduler": lr_scheduler,
                "interval": "batch",
                "frequency": 1,
            },
        }
```

This then advances on every batch step instead of every optimizer step

### Alternatives

# Alternative: Inject batch number

The best alternative I found is that users may override `lr_scheduler_step()` in LightningModule and manually track batch indices on the class, then inject them into the schedule:

```python
    def on_train_batch_start(self, batch, batch_idx):
        self.batch_num = batch_idx
    
    def lr_scheduler_step(self, scheduler, metric):
        scheduler.step(self.batch_num)
```

This is unintuitive, breaks Lightning's declarative config pattern, and requires understanding Lightning's internal stepping mechanism. Lightning should support batch stepping as part of auto training instead. 

### Additional context

Feel free to consult this article for an example of a process where it would have been useful (jump down to the LightningModule Integration section): [Article](https://medium.com/@chris.oquinn2/logical-batch-size-scheduling-in-lighting-cbd383c1ca09)

I would be happy to try implementing it myself if anyone is interested, but would like to know testing requirements and any dependencies. I have preliminarily identified "TrainEpochLoop" as the class in need of changes. 

cc @lantiga

**Comment by deependujha** [comment]
Thanks for the detailed proposal and the write-up,the logical-batch-size example is a nice illustration of the underlying need.

After looking into it, I don't think we should add `interval: "batch"` to core, because the behavior is already fully supported through manual optimization, and adding it would introduce semantics that are genuinely ambiguous under gradient accumulation.

---

### Why not core

The `"batch"` clock and the optimizer-step clock only diverge when `accumulate_grad_batches > 1`. In that regime, a per-batch scheduler would advance multiple times between weight updates. For stateless schedulers that's harmless, but for `ReduceLROnPlateau` and other stateful/metric-driven schedulers it's incoherent, the metric isn't even refreshed on non-step batches. Supporting `"batch"` in core would mean either carving those schedulers out or defining ad-hoc behavior for them, plus extending the `lr_scheduler_config` validation/JSON-schema and docs to carry a third interval permanently. That's a lasting maintenance and semantics cost for a case that already has a clean answer.

### The supported path

For manual optimization, the LR Scheduling docs already cover this directly: https://lightning.ai/docs/pytorch/stable/model/manual_optimization.html


<img width="1464" height="772" alt="Image" src="https://github.com/user-attachments/assets/41fcfc83-ca25-4f6d-b50b-94ec858fc909" />

---

### One correctness note on the workaround

The alternative in the issue (and in the linked article) keys the schedule off `batch_idx` via `on_train_batch_start`. `batch_idx` resets to 0 every epoch, so that formulation is only valid for single-epoch runs, which the article correctly scopes itself to. If anyone wants this to generalize past one epoch, drive the schedule off a monotonic counter (`trainer.global_step`, or your own counter incremented per batch if you specifically need per-forward-batch rather than per-optimizer-step granularity) rather than `batch_idx`.

---

So: the per-batch-LR use case is real and already achievable today, via documented manual optimization, or a small callback, without adding permanent config surface to core.

Going to close as `won't fix` on that basis, but happy to reopen if there's a concrete scenario neither path covers. Thanks again for the thorough issue.
